src/fd/eventfd.rs

```diff
@@ -0,0 +1,202 @@
+use alloc::boxed::Box;
+use alloc::collections::vec_deque::VecDeque;
+use core::future::{self, Future};
+use core::mem;
+use core::task::{Poll, Waker};
+
+use async_lock::Mutex;
+use async_trait::async_trait;
+
+use crate::fd::{block_on, EventFlags, IoError, ObjectInterface, PollEvent};
+
+#[derive(Debug)]
+struct EventState {
+	pub counter: u64,
+	pub read_queue: VecDeque<Waker>,
+	pub write_queue: VecDeque<Waker>,
+}
+
+impl EventState {
+	pub fn new(counter: u64) -> Self {
+		Self {
+			counter,
+			read_queue: VecDeque::new(),
+			write_queue: VecDeque::new(),
+		}
+	}
+}
+
+#[derive(Debug)]
+pub(crate) struct EventFd {
+	state: Mutex<EventState>,
+	flags: EventFlags,
+}
+
+impl Clone for EventFd {
+	fn clone(&self) -> Self {
+		let counter = block_on(async { Ok(self.state.lock().await.counter) }, None).unwrap();
+		Self {
+			state: Mutex::new(EventState::new(counter)),
+			flags: self.flags,
+		}
+	}
+}
+
+impl EventFd {
+	pub fn new(initval: u64, flags: EventFlags) -> Self {
+		debug!("Create EventFd {}, {:?}", initval, flags);
+		Self {
+			state: Mutex::new(EventState::new(initval)),
+			flags,
+		}
+	}
+}
+
+#[async_trait]
+impl ObjectInterface for EventFd {
+	async fn async_read(&self, buf: &mut [u8]) -> Result<usize, IoError> {
+		let len = mem::size_of::<u64>();
+
+		if buf.len() < len {
+			return Err(IoError::EINVAL);
+		}
+
+		future::poll_fn(|cx| {
+			if self.flags.contains(EventFlags::EFD_SEMAPHORE) {
+				let mut pinned = core::pin::pin!(self.state.lock());
+				if let Poll::Ready(mut guard) = pinned.as_mut().poll(cx) {
+					if guard.counter > 0 {
+						guard.counter -= 1;
+						let tmp = u64::to_ne_bytes(1);
+						buf[..len].copy_from_slice(&tmp);
+						if let Some(cx) = guard.write_queue.pop_front() {
+							cx.wake_by_ref();
+						}
+						Poll::Ready(Ok(len))
+					} else {
+						guard.read_queue.push_back(cx.waker().clone());
+						Poll::Pending
+					}
+				} else {
+					Poll::Pending
+				}
```

Comment:
We can extract `Poll::Ready` with [`core::task::ready`](https://doc.rust-lang.org/stable/core/task/macro.ready.html):

```suggestion
				let mut guard = ready!(pinned.as_mut().poll(cx));
				if guard.counter > 0 {
					guard.counter -= 1;
					let tmp = u64::to_ne_bytes(1);
					buf[..len].copy_from_slice(&tmp);
					if let Some(cx) = guard.write_queue.pop_front() {
						cx.wake_by_ref();
					}
					Poll::Ready(Ok(len))
				} else {
					guard.read_queue.push_back(cx.waker().clone());
					Poll::Pending
				}
```

If you like, I can do this after merging this PR or change this PR, however you prefer.